src/http/fetch.rs

```diff
@@ -163,31 +102,117 @@ pub(crate) fn init(ctx: &Ctx<'_>, globals: &Object) -> Result<()> {
-        return (Some(url_string), None);
-    } else if resource.is_object() {
-        let resource_obj = resource.into_object().unwrap();
-        return (None, Some(resource_obj));
+struct FetchOptions {
+    method: hyper::Method,
+    url: String,
+    headers: Option<Headers>,
+    body: Full<Bytes>,
+}
+
+fn get_fetch_options<'js>(
+    ctx: &Ctx<'js>,
+    resource: Value<'js>,
+    opts: Opt<Value<'js>>,
+) -> Result<FetchOptions> {
+    let mut url = None;
+    let mut resource_opts = None;
+    let mut arg_opts = None;
+    let mut headers = None;
+    let mut method = None;
+    let mut body = None;
+    if let Some(obj) = resource.as_object() {
+        let obj = obj.clone();
+        if obj.instance_of::<crate::http::Request>() {
+            resource_opts = Some(obj);
+        } else if let Some(to_string) = obj.get::<_, Option<Function>>(PredefinedAtom::ToString)? {
+            url = Some(to_string.call::<_, String>((obj,))?);
+        } else {
+            resource_opts = Some(obj);
+        }
+    } else {
+        url = Some(resource.get::<Coerced<String>>()?.to_string());
+    }
+
+    if let Some(options) = opts.0 {
+        arg_opts = options.into_object();
+    }
+
+    if resource_opts.is_some() || arg_opts.is_some() {
+        if let Some(method_opt) = get_option::<String>("method", &arg_opts, &resource_opts)? {
+            method = Some(match method_opt.as_str() {
+                "GET" => Ok(hyper::Method::GET),
+                "POST" => Ok(hyper::Method::POST),
+                "PUT" => Ok(hyper::Method::PUT),
+                "CONNECT" => Ok(hyper::Method::CONNECT),
+                "HEAD" => Ok(hyper::Method::HEAD),
+                "PATCH" => Ok(hyper::Method::PATCH),
+                "DELETE" => Ok(hyper::Method::DELETE),
+                _ => Err(Exception::throw_type(
+                    ctx,
+                    &format!("Invalid HTTP method: {}", method_opt),
+                )),
+            }?);
+        }
+
+        if let Some(body_opt) = get_option::<Value>("body", &arg_opts, &resource_opts)? {
+            let bytes = get_bytes(ctx, body_opt)?;
+            body = Some(Full::from(bytes));
+        }
+
+        if let Some(url_opt) = get_option::<String>("url", &arg_opts, &resource_opts)? {
+            url = Some(url_opt);
+        }
+
+        if let Some(headers_op) = get_option::<Value>("headers", &arg_opts, &resource_opts)? {
+            headers = Some(Headers::from_value(ctx, headers_op)?);
+        }
     }
-    (None, None)
+
+    let url = match url {
+        Some(url) => url,
+        None => return Err(Exception::throw_reference(ctx, "Missing required url")),
+    };
+
+    Ok(FetchOptions {
+        method: method.unwrap_or_default(),
+        url,
+        headers,
+        body: body.unwrap_or_default(),
+    })
+}
+
+fn get_option<'js, V: FromJs<'js> + Sized>(
+    arg: &str,
+    a: &Option<Object<'js>>,
+    b: &Option<Object<'js>>,
+) -> Result<Option<V>> {
```

Comment:
It is a minor detail, but I would use `resource` and `options` for readability.

```javascript
fetch(resource)
fetch(resource, options)
```

Reference:
https://developer.mozilla.org/en-US/docs/Web/API/fetch#syntax